### PR TITLE
Add a `resource.update` convenience function

### PR DESCRIPTION
**Pull request by negz**
### Description of your changes

This function can be used to add or update a resource.

For example, to add a desired composed bucket (assuming the "bucket" resource doesn't yet exist):

```python
resource.update(rsp.desired.resources["bucket"], bucket)
```

Or to update the desired XR:

```python
resource.update(rsp.desired.composite.resource, partial_xr)
```

The source resource can be a dictionary, a protobuf struct, or a Pydantic model.

I have:

- [x] Read and followed Crossplane's [contribution process].
- [x] Added or updated unit tests for my change.

[contribution process]: https://git.io/fj2m9
[docs]: https://docs.crossplane.io/contribute/contribute

### CODE REVIEW

File: crossplane/function/resource.py [hunk 1]
```diff
@@ -17,12 +17,41 @@
...
+
+    Use update to add or update the supplied resource. If the resource doesn't
+    exist, it'll be added. If the resource does exist, it'll be updated. The
+    update method semantics are the same as a dictionary's update method. Fields
+    that don't exist will be added. Fields that exist will be overwritten.
+
+    The source can be a dictionary, a protobuf Struct, or a Pydantic model.
+    """
+    # If the resource has a model_dump attribute it's a Pydantic BaseModel.
+    if hasattr(source, "model_dump"):
```
Comment (negz) [inline crossplane/function/resource.py:40]: @bobh66 Let me know if there's a more Pythonic way to determine the type of the `source` argument. I'm still pretty rusty. 😄

File: tests/testdata/models/io/k8s/apimachinery/pkg/apis/meta/v1.py [hunk 2]
```diff
@@ -0,0 +1,301 @@
+# generated by datamodel-codegen:
```
Comment (negz) [inline tests/testdata/models/io/k8s/apimachinery/pkg/apis/meta/v1.py:1]: This is all generated code - no need to review it.

File: pyproject.toml [hunk 3]
```diff
@@ -16,7 +16,13 @@ classifiers = [
   "Programming Language :: Python :: 3.11",
 ]
 
-dependencies = ["grpcio==1.*", "grpcio-reflection==1.*", "protobuf==5.27.2", "structlog==24.*"]
+dependencies = [
+  "grpcio==1.*",
+  "grpcio-reflection==1.*",
+  "protobuf==5.27.2",
+  "pydantic==2.*",
```
Comment (negz) [inline pyproject.toml:23]: I don't love introducing a dependency on `pydantic` just for the type annotation really. I think if it proves to be a pain we could move it out to an optional dependency.

File: crossplane/function/resource.py (hunk 1 above)
Comment (RedbackThomson) [reply crossplane/function/resource.py:40]: I believe `isinstance` would be how to determine if it's a subclass

Comment (negz) [reply crossplane/function/resource.py:40]: @RedbackThomson Oh duh, good point.

Comment (negz) [reply crossplane/function/resource.py]: I knew Python recently got a `match` statement which is kinda like a Go switch. A bit of Googling led me to find that they now have a 'class match', which is similar to a Go type switch. I've updated the code to use that.

**Pull request merged**